### ISSUE TITLE
questions about logic of work

### ISSUE THREAD

**Issue by ahmetozalp** [peer-calls/peer-calls#93]
I want to list all the streams here. How can I do it?** https://github.com/jeremija/peer-calls/blob/master/src/client/components/Media.tsx#L67

**Comment by jeremija** [comment]
Does this help?

```diff
diff --git a/src/client/components/Media.tsx b/src/client/components/Media.tsx
index 1ce42ec..e2292c1 100644
--- a/src/client/components/Media.tsx
+++ b/src/client/components/Media.tsx
@@ -56,11 +56,14 @@ export class MediaForm extends React.PureComponent<MediaProps> {
     event.preventDefault()
     const { props } = this
     const { audio, video } = props
+    let localStream: MediaStream | undefined
     try {
-      await props.getMediaStream({ audio, video })
+      const payload = await props.getMediaStream({ audio, video })
+      localStream = payload.stream
     } catch (err) {
       props.logError('Error getting media stream: {0}', err)
     }
+    console.log(localStream)

     props.logInfo('Dialling...')
     try {
```

**Comment by ahmetozalp** [comment]
no. I want to get the list of users in the room. I want to do, to check if the room has the same userID

**Comment by ahmetozalp** [comment]
`MediaStream {id: "cJmyaLUxm654tSXSR8h88J5TVO5Ch00UxddS", active: true, onaddtrack: null, onremovetrack: null, onactive: null, …}
id: "cJmyaLUxm654tSXSR8h88J5TVO5Ch00UxddS"
active: true
onaddtrack: null
onremovetrack: null
onactive: null
oninactive: null
__proto__: MediaStream`

**Comment by ahmetozalp** [comment]
I couldn't throw it from props here, it gave error all the time. There must be a simpler way to get this list.

**Comment by jeremija** [comment]
```ts
diff --git a/src/client/actions/CallActions.ts b/src/client/actions/CallActions.ts
index e5dd8bf..28ac3dc 100644
--- a/src/client/actions/CallActions.ts
+++ b/src/client/actions/CallActions.ts
@@ -5,6 +5,7 @@ import store, { ThunkResult } from '../store'
 import { callId, userId } from '../window'
 import * as NotifyActions from './NotifyActions'
 import * as SocketActions from './SocketActions'
+import { SocketEvent } from '../../shared'

 export interface InitAction {
   type: 'INIT'
@@ -34,14 +35,14 @@ export const init = (): ThunkResult<Promise<void>> => async dispatch => {

 export const dial = makeAction(
   DIAL,
-  () => new Promise<void>((resolve, reject) => {
+  () => new Promise<SocketEvent['users']>((resolve, reject) => {
     SocketActions.handshake({
       socket,
       roomName: callId,
       userId,
       store,
     })
-    socket.once(SOCKET_EVENT_USERS, () => resolve())
+    socket.once(SOCKET_EVENT_USERS, payload => resolve(payload))
     setTimeout(reject, 10000, new Error('Dial timed out!'))
   }),
 )

```

This should allow you to get a list of users after calling `dial()`:

```
const payload = await props.dial()
```

**Comment by ahmetozalp** [comment]
`
props.logInfo('Konferans Kontrol Ediliyor...')
const result = payload.users.filter(user => user.userId === userId)
if(result.length > 1){
        const error = { message: 'Aynı kimlik de odada kullanıcı mevcut' };
        throw error.message;
}
`

thank you. only left the room left user

**Comment by jeremija** [comment]
Since the current user is joining the room, it is expected that the current user is in the `users` list.

 You can filter the current user out:

```typescript
const otherUsers = payload.users.filter(user => user.userId !== userId)
```

**Comment by ahmetozalp** [comment]
Thank you 2side works for me.

I will have 1 error and 1 question for you.
question;
1- What is redis used for? I don't use it and I didn't encounter an error.

error;
1- How can I delete the 'mirroring' class on the back camera works properly on the front camera but it should not work on the back camera

**Comment by jeremija** [comment]
I got an e-mail notification about a question on how to stop the mirroring of the back camera.

Theoretically, you could enable mirroring only when the `label` of a `MediaStreamTrack` with  `kind === 'video'` contains `front`:

```typescript
const shouldMirror = !!localMediaStream.getVideoTracks()[0]?.label.indexOf('front')
```

This works in Chrome. However, it doesn't seem to work correctly in all browsers so I've refrained from using it. For example, the `MediaStreamTrack.label` on Firefox for Android Nightly does not anything useful to determine the camera orientation.

The bottom line is: the picture will only be mirrored to the local user recording the video. Other users will see the picture normally.

**Comment by ahmetozalp** [comment]
I could not bring the device list to [this](https://github.com/peer-calls/peer-calls/blob/master/src/client/components/Video.tsx#L67) line

**Comment by jeremija** [comment]
You should be able to modify the `mapStateToProps` function in `containers/App.tsx` and pass the `state.media.devices` to `components/App.tsx`. From there, pass the device list to `components/Videos.tsx` and to `components/Video.tsx`. Note that the interfaces of all the affected components will have to be modified to accept a new prop.

Alternatively, you could make the `Video.tsx` a redux-connected component and create a separate `mapStateToProps` there. See the redux documentation for `connect` function.

**Comment by ahmetozalp** [comment]
Hello,

Thank you very much for your answer. I never needed redux because I used the context structure, but now I will learn redux

**Comment by ahmetozalp** [comment]
```Javascript
let mirror = mirrored
    const videoDevices = devices.filter(device => device.type === "videoinput")
    if(videoDevices.length > 1 && mirrored){
      if (stream?.stream.getVideoTracks()[0]?.label.match(/(back|Back|arka|Arka)/) ) {
        mirror = false
      }
    }

    const className = classnames('video-container', { active, mirrored: mirror })
```

**Comment by ahmetozalp** [comment]
@5 step:

```Javascript
console.log(payload.users)
```
![Image description](https://i.hizliresim.com/0jioCx.png)

How do I add the 3rd object to this field?

**Comment by jeremija** [comment]
https://github.com/peer-calls/peer-calls/blob/master/src/shared/SocketEvent.ts#L4-L7

Since this project uses TypeScript, I'd recommend you to use a text editor with support for [`tsserver`](https://github.com/Microsoft/TypeScript/wiki/Standalone-Server-%28tsserver%29), if you don't already use one. It will help you a lot with finding references in cases like these.

**Comment by ahmetozalp** [comment]
I saw here, I added the line, but I think I have to trigger it from somewhere.



i am using typescript my editor vscode in react-native projects

**Comment by jeremija** [comment]
You should get a compile error if you add a required field that is not populated.

**Comment by ahmetozalp** [comment]
I added the area you said but it didn't give an error and it works as if there were none

**Comment by ahmetozalp** [comment]
fix.

https://github.com/peer-calls/peer-calls/blob/master/src/server/socket.ts
